### DISCUSSION TITLE
What is your opinion about umlauts (ä, ö, ü) in the URL?

### DISCUSSION THREAD

**Discussion by PetersOtto** [datenstrom/community#774] [Ask a question] [answered]
Hello Community!
I realized that the Blog extension does not convert the umlauts (ä, ö, ü) in the domain automatically. I feel not good with that. Is that no longer a problem?
I know it is possible to use TitleSlug ( https://github.com/datenstrom/yellow/discussions/326) to customize the URL

```
Title: Blogseite an der Küste
TitleSlug: Blogseite an der Kueste
Published: 2018-05-27 12:00:00
Template: blog
```

However, it is not so easy to explain that a customer.
Better would be a automatically change from ä to ae.
What is your opinion?

![YellowUmlaute](https://user-images.githubusercontent.com/52236754/161008913-cfbc2e8c-69e9-436a-a660-112686795f5d.png)

**Comment by schulle4u** [comment]
Most web servers and browsers shouldn't have a problem parsing umlauts nowadays, however as an internet child of the late 90s and early 2000s I personally prefer clean URLs in my projects too. 🙂 Although I would vote for a better URL normalisation, I don't see a problem in manually setting the `TitleSlug`. In fact there are even more title settings to control how page titles are displayed throughout the website, the slug is just one of them. 

Another approach to automate URL creation could be to use placeholders and patterns for the `TitleSlug` setting. Here is an example `content/shared/page-new-blog.md` which uses the entries date as part of the URL. 

```
---
Title: Blog page
TitleSlug: blog-@date
Published: @datetime
Author: @username
Layout: blog
Tag: Example
---
This is a new blog page.
```

Now the `TitleSlug` will be prefilled with the publication date and a text pattern upon page creation. This is just an example, the possibilities are nearly endless.

**Reply by glow12121** [reply]
> Most web servers and browsers shouldn't have a problem parsing umlauts nowadays, however as an internet child of the late 90s and early 2000s I personally prefer clean URLs in my projects too.

I agree. Latin letters with accents and umlauts are a bad idea for URLs, just like Cyrillic, Arabic, Asian, or any other alphabets, other than plain English.

**Comment by PetersOtto** [comment]
Oh, yes! The possibility to work with placeholders was not in my mind.  Thank you! 
However I think for SEO would it good to have the Title in the domain.
Do you know if shortcodes would be possible in the Header (TitleSlug)?
Something like that:

```
Title: Blogseite an der Küste
TitleSlug: [ChangeTitleUmlauts]
Published: 2018-05-27 12:00:00
Template: blog
```

**Reply by schulle4u** [reply]
This is not possible, but here's another solution you could build on. Have a look at datenstrom/community#372.

**Reply by PetersOtto** [reply]
Oh, great! Now I have enough solutions :) Thx!

**Reply by GiovanniSalmeri** [reply]
If you prefer a built-in function (which could be not installed in your version of PHP!), see https://www.php.net/manual/en/function.iconv.php

An example of use (and important remark) is in comment https://www.php.net/manual/en/function.iconv.php#74101

**Reply by PetersOtto** [reply]
Thank you @GiovanniSalmeri, I will have a closer look to it!

**Reply by markseuffert** [reply]
Maybe someone wants to improve on datenstrom/community#372, add more languages, [make and publish](https://github.com/annaesvensson/yellow-publish) a transliterate extension? 

All it needs now is someone to volunteer. 😀

**Reply by PetersOtto** [reply]
I would do/try that but I think my skills in php, yellow API and github are a lot to low...
I will work me through your documentation about "helping you" and come back to you.

**Comment by PetersOtto** [comment, marked as answer]
This works fine for me (VERSION = "0.8.19"):
1. put the `TitleSlug:` in the Header and let it empty.

```
Title: Blogseite an der Küste
TitleSlug:
Published: 2018-05-27 12:00:00
Template: blog
```
2. Create a `umlaute.php` in the extension folder with the following code. Have a look here: https://github.com/datenstrom/yellow/discussions/372 Thank you @schulle4u and @markseuffert 

```
<?php

class YellowUmlaute {
    const VERSION = "0.8.19";
    public $yellow;         //access to API

    // Handle initialisation
    public function onLoad($yellow) {
        $this->yellow = $yellow;
    }
    
    // Handle content file editing
    public function onEditContentFile($page, $action, $email) {
        if (!$page->isExisting("titleSlug")) {
            $replaceData = $this->getReplaceData();
            $titleSlug = str_replace(array_keys($replaceData), array_values($replaceData), $page->get("title"));
            $titleSlug = preg_replace("/[^\pL\d\-\_\ ]/u", "-", $titleSlug);
            $page->rawData = $this->yellow->toolbox->setMetaData($page->rawData, "titleSlug", $titleSlug);
        }
    }
    
    // Return text replace data, UTF8 to ASCII
    public function getReplaceData() {
        return array(
            "Ä" => "AE", "ä" => "ae", "Ü" => "UE", "ü" => "ue","Ö" => "OE", "ö" => "oe",
        );
    }
}
```

**Reply by IngoBelka** [reply]
Only a thought, what about to add "ß" =>"sz".

**Reply by schulle4u** [reply]
Ss is the common way to rewrite it.

**Reply by IngoBelka** [reply]
Sorry for the little joke. ;-)
